### PR TITLE
optional link, setup regex, on off link

### PR DESCRIPTION
**Pull request by devmario**
Test fail fix

### CODE REVIEW

File: ActiveLabel.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -287,8 +290,11 @@
 			isa = PBXSourcesBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				BC6FFC1F1CCABF2E001C195F /* ActiveLabel.swift in Sources */,
 				8F0249C51B998A66005D8035 /* ViewController.swift in Sources */,
 				8F0249C31B998A66005D8035 /* AppDelegate.swift in Sources */,
+				BC6FFC1E1CCABF1C001C195F /* RegexParser.swift in Sources */,
+				BC6FFC201CCABF31001C195F /* ActiveType.swift in Sources */,
```
Comment (poolqf) [inline ActiveLabel.xcodeproj/project.pbxproj]: The demo should not have this files

File: ActiveLabel/ActiveLabel.swift [hunk 2]
```diff
@@ -52,6 +58,11 @@ public protocol ActiveLabelDelegate: class {
     public func handleURLTap(handler: (NSURL) -> ()) {
         urlTapHandler = handler
     }
+    
+    public func handleOptionalTab(pattern:String, handler: (String) -> ()) {
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: Please follow the style guide `pattern: String`.

Also, `Tap`

File: ActiveLabel/ActiveLabel.swift [hunk 3]
```diff
@@ -166,6 +178,9 @@ public protocol ActiveLabelDelegate: class {
     private var mentionTapHandler: ((String) -> ())?
     private var hashtagTapHandler: ((String) -> ())?
     private var urlTapHandler: ((NSURL) -> ())?
+
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: Remove space

File: ActiveLabel/ActiveLabel.swift [hunk 4]
```diff
@@ -36,6 +36,12 @@ public protocol ActiveLabelDelegate: class {
     @IBInspectable public var URLSelectedColor: UIColor? {
         didSet { updateTextStorage(parseText: false) }
     }
+    @IBInspectable public var optionalColor: UIColor = .blueColor() {
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: Remove default color

File: ActiveLabel/ActiveLabel.swift [hunk 5]
```diff
@@ -166,6 +178,9 @@ public protocol ActiveLabelDelegate: class {
     private var mentionTapHandler: ((String) -> ())?
     private var hashtagTapHandler: ((String) -> ())?
     private var urlTapHandler: ((NSURL) -> ())?
+    
+    private var optionalPatterns:[String:String] = [:]
+    private var optionalTapHandler:[String:((String) -> ())?] = [:]
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]:
```swift
private var optionalPatterns: [String : String] = [:]
private var optionalTapHandler: [String: ((String) -> ())?] = [:]
```

File: ActiveLabel/ActiveLabel.swift [hunk 6]
```diff
@@ -379,6 +403,14 @@ public protocol ActiveLabelDelegate: class {
         }
         urlHandler(url)
     }
+    
+    private func didTapOptional(pattern:String, optional: String) {
+        guard let handler = optionalTapHandler[pattern] else {
+            delegate?.didSelectText(optional, type: .Optional)
+            return
+        }
+        handler!(optional)
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: `handler?(optional)`

File: ActiveLabel/RegexParser.swift [hunk 7]
```diff
@@ -33,4 +33,11 @@ struct RegexParser {
         return urlDetector.matchesInString(text, options: [], range: range)
     }
     
+    static func getOptionals(fromText text: String, pattern:String, range: NSRange) -> [NSTextCheckingResult] {
+        if let match = try? NSRegularExpression(pattern: pattern, options: [.CaseInsensitive]).matchesInString(text, options: [], range: range) {
+            return match
+        }
+        return []
```
Comment (poolqf) [inline ActiveLabel/RegexParser.swift]: Please do it consistently with previous methods

File: ActiveLabel/ActiveLabel.swift [hunk 8]
```diff
@@ -15,6 +15,13 @@ public protocol ActiveLabelDelegate: class {
 
 @IBDesignable public class ActiveLabel: UILabel {
     
+    public var on:[ActiveType:Bool] = [
+        .Mention : true,
+        .Hashtag: true,
+        .URL: true,
+        .Optional: true,
+    ]
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: This should be all of them different properties, not a dict

File: ActiveLabel/ActiveLabel.swift [hunk 9]
```diff
@@ -268,22 +275,30 @@ public protocol ActiveLabelDelegate: class {
...
-        activeElements[.Hashtag]?.appendContentsOf(hashtagElements)
+        if (on[.Hashtag] != nil && on[.Hashtag]!) {
+            let hashtagElements = ActiveBuilder.createHashtagElements(fromText: textString, range: textRange, filterPredicate: hashtagFilterPredicate)
+            activeElements[.Hashtag]?.appendContentsOf(hashtagElements)
+        }
         
         //MENTIONS
-        let mentionElements = ActiveBuilder.createMentionElements(fromText: textString, range: textRange, filterPredicate: mentionFilterPredicate)
-        activeElements[.Mention]?.appendContentsOf(mentionElements)
+        if (on[.Mention] != nil && on[.Mention]!) {
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: All of those `if`s should rely on properties

File: ActiveLabel/ActiveLabel.swift [hunk 10]
```diff
@@ -15,6 +15,18 @@ public protocol ActiveLabelDelegate: class {
...
+        RegexParser.mentionRegex = regex
+    }
+    
+    static func setupHashtagRegex(regex:NSRegularExpression) {
+        RegexParser.hashtagRegex = regex
+    }
+    
+    static func setupURLRegex(regex:NSRegularExpression) {
+        RegexParser.urlDetector = regex
+    }
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: Move all those methods below `customize()` one (https://github.com/optonaut/ActiveLabel.swift/pull/72/commits/bb7f7a9d1fa337dec115e193ed2ec4d37c05b023#diff-a848e4d55390095e97192f534b60c1d0R149). Also `regex: NSRegularExpression ` and make them public

File: ActiveLabelDemo/ViewController.swift [hunk 11]
```diff
@@ -17,18 +17,24 @@ class ViewController: UIViewController {
         super.viewDidLoad()
         
         label.customize { label in
-            label.text = "This is a post with #multiple #hashtags and a @userhandle. Links are also supported like this one: http://optonaut.co."
+            
+            label.text = "This is a post with #multiple and a @userhandle. change pattern @+userhandle.Links are also supported like this one: http://optonaut.co. patern link 'Optional'"
+            
+            ActiveLabel.setupMentionRegex(try! NSRegularExpression(pattern: "(?:^|\\s|$|[.])@[\\p{L}\\+0-9_]*", options: [.CaseInsensitive]))
```
Comment (poolqf) [inline ActiveLabelDemo/ViewController.swift]: Remove this line

**Comment by poolqf** [comment]
Hi @devmario, thanks for this PR. I like a lot the idea.

There are just 2 things:
- The `optional` naming is super misleading. Could you please rename all `optional` to `custom`, makes more sense
- Please try to follow the same code style as the library has

Another time, thanks a lot for the PR

File: ActiveLabel/ActiveLabel.swift [hunk 12]
```diff
@@ -15,6 +15,25 @@ public protocol ActiveLabelDelegate: class {
...
+        RegexParser.mentionRegex = regex
+    }
+    
+    public static func setupHashtagRegex(regex:NSRegularExpression) {
+        RegexParser.hashtagRegex = regex
+    }
+    
+    public static func setupURLRegex(regex:NSRegularExpression) {
+        RegexParser.urlDetector = regex
+    }
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: Move all those methods below `customize()` one (https://github.com/optonaut/ActiveLabel.swift/pull/72/commits/bb7f7a9d1fa337dec115e193ed2ec4d37c05b023#diff-a848e4d55390095e97192f534b60c1d0R149).

File: ActiveLabel/ActiveLabel.swift [hunk 13]
```diff
@@ -251,17 +286,31 @@ public protocol ActiveLabelDelegate: class {
...
-
+        if (on[.Hashtag] != nil && on[.Hashtag]!) {
+            let hashtagElements = ActiveBuilder.createHashtagElements(fromText: textString, range: textRange, filterPredicate: hashtagFilterPredicate)
+            activeElements[.Hashtag]?.appendContentsOf(hashtagElements)
+        }
+        
         //MENTIONS
-        let mentionElements = ActiveBuilder.createMentionElements(fromText: textString, range: textRange, filterPredicate: mentionFilterPredicate)
-        activeElements[.Mention]?.appendContentsOf(mentionElements)
+        if (on[.Mention] != nil && on[.Mention]!) {
```
Comment (poolqf) [inline ActiveLabel/ActiveLabel.swift]: All of those `if`s should rely on properties

**Comment by poolqf** [comment]
@devmario are you going to follow with this PR?

**Comment by poolqf** [comment]
Hi @devmario, are you going to fix this PR or should I close it?

**Comment by devmario** [comment]
Yes @poolqf . 
I'm join #78. Sorry I'm late

**Comment by devmario** [comment]
Wait. will I fixed it

**Comment by devmario** [comment]
Hey @poolqf ~~ ;)
I'm all Fixed~!
Check plz~

**Comment by devmario** [comment]
@poolqf All fixed. check my PR plz

**Comment by devmario** [comment]
Move setup regex at below customize

File: ActiveLabel/ActiveLabel.swift (hunk 12 above)
Comment (devmario) [inline ActiveLabel/ActiveLabel.swift]: All Your idea fixed.
Moved complete all setup Regex Methods at below customize

**Comment by poolqf** [comment]
With today's merge, this PR is now not mergeable anymore. The new implementation is allowing adding new default types of highlighting patterns easily. You should not have any problem to adapt this PR to be merged.

**Pull request closed without merging**